### PR TITLE
Add bug report modal

### PR DESCRIPTION
**Pull request by EpokK**
### Added

* useConfig (React hook)
* ConfigContext
* BugModal
* FileInput

### Screenshots

Capture d’écran 2019-04-12 à 08 34 14
Capture d’écran 2019-04-12 à 08 34 24

### Note

- We hide by default system information (browser, os, etc...)
- We could add a `<select>` to set a title for the issue (customer support needs to provide the list of options)

### CODE REVIEW

File: containers/support/AttachScreenshot.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                await Promise.all(
+                    images.map((img) =>
+                        resize(img, MAX_SIZE_SCREENSHOT).then((base64str) => ({
+                            name: img.name,
+                            blob: toBlob(base64str)
+                        }))
+                    )
+                )
+            );
+        } else {
```
Comment (dhoko) [inline containers/support/AttachScreenshot.js]: use a return first ;) 
if no images -> error

File: containers/support/AttachScreenshot.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                text: c('Error notification in the bug report modal when the user upload file').t`No image selected`
+            });
+        }
+    };
+
+    if (attached) {
+        return (
+            <>
+                <Text>
+                    <Icon name="insert-image" /> {c('Info').t`Screenshot(s) attached`}
```
Comment (dhoko) [inline containers/support/AttachScreenshot.js]: Maybe we can display a small preview (~ 500*400) ?

File: containers/support/BugModal.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    const os = getOS();
+    const browser = getBrowser();
+    const device = getDevice();
+    const [model, update] = useState({
+        OS: os.name,
+        OSVersion: os.version || '',
+        Browser: browser.name,
+        BrowserVersion: browser.version,
+        Resolution: `${window.innerHeight} x ${window.innerWidth}`,
+        Client: 'Angular', // TODO define a new one with the API
```
Comment (dhoko) [inline containers/support/BugModal.js]: We should pass the config from the app as a prop to the modal, so we can extend it with specific keys

File: containers/support/BugModal.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        ClientVersion: '3.15.23', // TODO get it from app
+        ClientType: 1, // TODO get it from config
+        DeviceName: device.vendor,
+        DeviceModel: device.model,
+        Title: `[ProtonMail Settings React] Bug [${location.path}]`,
```
Comment (dhoko) [inline containers/support/BugModal.js]: Idem for the title, it should comes as a prop --> `ProtonMail Settings React`

File: containers/support/BugModal.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    const { request, loading } = useApiWithoutResult(reportBug);
+    const link = (
+        <Href url="https://protonmail.com/support/knowledge-base/how-to-clean-cache-and-cookies/">{c('Link')
+            .t`clearing your browser cache`}</Href>
+    );
+    const handleChange = (key) => ({ target }) => update({ ...model, [key]: target.value });
+
+    const handleReset = () => {
+        // eslint-disable-next-line no-unused-vars
+        const { Attachments, ...newModel } = model;
```
Comment (dhoko) [inline containers/support/BugModal.js]: Can you create a omit function or import the one from lodash ?

File: containers/support/BugModal.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        update(newModel);
+    };
+
+    const handleUpload = (images) => update({ ...model, Attachments: images });
+
+    const format = (parameters = {}) => {
```
Comment (dhoko) [inline containers/support/BugModal.js]: doc ?

File: containers/support/BugModal.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        return Object.entries(parameters).reduce((acc, [key, value]) => {
```
Comment (dhoko) [inline containers/support/BugModal.js]: Object.values().reduce(acc, key) => {
const value = parameters[key
}

File: containers/support/BugModal.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+            if (value instanceof Array) {
```
Comment (dhoko) [inline containers/support/BugModal.js]: Why not Array.isArray ? Do we have a specific use case ?

File: containers/support/AttachScreenshot.js (hunk 2 above)
Comment (EpokK) [reply containers/support/AttachScreenshot.js]: I let design team design.

File: containers/support/BugModal.js (hunk 3 above)
Comment (EpokK) [reply containers/support/BugModal.js]: A `Context` is probably more appropriate.

File: containers/support/BugModal.js (hunk 4 above)
Comment (EpokK) [reply containers/support/BugModal.js]: You are right.

File: containers/support/BugModal.js (hunk 5 above)
Comment (EpokK) [reply containers/support/BugModal.js]: Good idea.

File: containers/support/BugModal.js (hunk 7 above)
Comment (EpokK) [reply containers/support/BugModal.js]: What's wrong with `Object.entries`?

File: containers/support/BugModal.js (hunk 8 above)
Comment (EpokK) [reply containers/support/BugModal.js]: I will change it, `instanceof` was previously used in the part of code I copied.

File: containers/support/BugModal.js [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        <Href url="https://protonmail.com/support/knowledge-base/how-to-clean-cache-and-cookies/">{c('Link')
+            .t`clearing your browser cache`}</Href>
+    );
+    const handleChange = (key) => ({ target }) => update({ ...model, [key]: target.value });
+
+    const getParameters = () => {
+        return images.reduce((acc, { name, blob }) => {
+            acc[name] = blob;
+            return acc;
+        }, model);
```
Comment (dhoko) [inline containers/support/BugModal.js]: :warning: side effect

File: containers/support/BugModal.js [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    const { CLIENT_ID, APP_VERSION } = useConfig();
+    const Client = CONFIGS[CLIENT_ID];
+    const [{ Name = '' }] = useUser();
+    const { createNotification } = useNotifications();
+    const [addresses = []] = useAddresses();
+    const [{ Email = '' } = {}] = addresses;
+    const os = getOS();
+    const browser = getBrowser();
+    const device = getDevice();
+    const [model, update] = useState({
```
Comment (dhoko) [inline containers/support/BugModal.js]: imho the defaut object should come from a helper, not as a default inside the modal

File: containers/support/BugModal.js [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    PrimaryButton,
+    useApiWithoutResult,
+    useUser,
+    useAddresses,
+    useNotifications,
+    useConfig
+} from 'react-components';
+import AttachScreenshot from './AttachScreenshot';
+
+const CONFIGS = {
```
Comment (dhoko) [inline containers/support/BugModal.js]: idem should be inside a helper

File: containers/support/AttachScreenshot.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        if (!images.length) {
+            return createNotification({
+                type: 'error',
+                text: c('Error notification in the bug report modal when the user upload file').t`No image selected`
+            });
+        }
+
+        setAttached(true);
+        onUpload(
+            await Promise.all(
```
Comment (dhoko) [inline containers/support/AttachScreenshot.js]: it should be a helper.

File: containers/support/AttachScreenshot.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    const [attached, setAttached] = useState(false);
+    const { createNotification } = useNotifications();
+
+    const handleClick = () => {
+        setAttached(false);
+        onReset();
+    };
+
+    const handleChange = async ({ target }) => {
+        const images = [...target.files].filter(({ type }) => /^image\//i.test(type));
```
Comment (dhoko) [inline containers/support/AttachScreenshot.js]: idem should be a hlper

File: containers/support/AttachScreenshot.js (hunk 12 above)
Comment (EpokK) [reply containers/support/AttachScreenshot.js]: I didn't get this one. It's a `Promise.all` running a helper function and formatting a specific result. Also, the logic is already extracted inside a helper.

File: containers/support/AttachScreenshot.js (hunk 13 above)
Comment (EpokK) [reply containers/support/AttachScreenshot.js]: This is a filter on a group of file to extract images. That's specific to this upload I think.

**Pull request merged**